File: erlang/ch1/ex122.erl
%%%-------------------------------------------------------------------
%%% @author Yohan Lee
%%% @copyright (C) 2019
%%% @doc
%%% cd("c:/Users/Yohan/IdeaProjects/sicp/erlang/ch1").
%%% compile with > c("exXXX.erl").
%%% if already compiled, load with > l(exXXX).
%%% run with > exXXX:main().
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(ex122).
-author("Yohan").

%% API
-export([main/0]).

main() ->
  io:fwrite("~w\n", [search_for_primes(1000, 1020)]),
  io:fwrite("~w\n", [search_for_primes(10000, 10040)]),
  io:fwrite("~w\n", [search_for_primes(100000, 100080)]).


smallest_divisor(N) -> find_divisor(N, 2).

find_divisor(N, Test) when Test * Test > N -> N;
find_divisor(N, Test) ->
  if
    N rem Test == 0 -> Test;
    true -> find_divisor(N, Test + 1)
  end.

prime(N) -> N == smallest_divisor(N).

report_prime(N, Prime) ->
  if
    Prime -> io:fwrite("~w ***\n", [N]);
    true -> io:fwrite("~w \n", [N])
  end.

prime_test(N) -> report_prime(N, prime(N)).

search_helper(N, End) ->
  if
    N =< End -> prime_test(N), search_helper(N + 2, End);
    true -> io:fwrite("finished\n")
  end.

search_for_primes(Start, End) ->
  if
    Start rem 2 == 1 -> search_helper(Start, End);
    true-> search_helper(Start + 1, End)
  end.

File: julia/ch1/Ex122.jl
"""
```jldoctest
julia> include("c:/Users/Yohan/IdeaProjects/sicp/julia/ch1/ExXXX.jl")

julia> ExXXX.func(arg)
```
"""
module Ex122

sq(n) = n * n

divides(a, b) = b % a == 0

function find_divisor(n, test)
    if sq(test) > n
        return n
    elseif divides(test, n)
        return test
    else
        return find_divisor(n, test + 1)
    end
end

smallest_divisor(n) = find_divisor(n, 2)

prime(n) = n == smallest_divisor(n)

function report_prime(n)
    if prime(n)
        println("$n ***")
    else
        println("$n")
    end
end

function search_helper(n, finish)
    if n <= finish
        report_prime(n)
        search_helper(n + 2, finish)
    else
        println("finished")
    end
end

search_for_primes(start, finish) = 
    search_helper(divides(2, start) ? start + 1 : start, finish)


end
